tiledb/sm/query/ast/test/unit-query-ast.cc

```diff
@@ -0,0 +1,356 @@
+  CHECK(
+      ast_node_to_str(combined_or4) ==
+      "(foo EQ 00 00 00 00 OR foo EQ 02 00 00 00 OR foo EQ 04 00 00 00 OR foo "
+      "EQ 06 00 00 00 OR foo EQ 08 00 00 00)");
+}
+
+TEST_CASE(
+    "ASTNode Constructors, complex tree with depth > 2",
+    "[QueryCondition][ast][constructor]") {
+  std::vector<int> vals = {1, 2, 3, 4, 5, 6, 7, 8, 9};
+  std::vector<tdb_unique_ptr<ASTNode>> ast_value_vector;
+  for (size_t i = 0; i < 7; ++i) {
+    auto node_val = tdb_unique_ptr<ASTNode>(tdb_new(
+        ASTNodeVal, "x", &vals[i], sizeof(vals[i]), QueryConditionOp::EQ));
+    CHECK(
+        ast_node_to_str(node_val) ==
+        "x EQ 0" + std::to_string(vals[i]) + " 00 00 00");
+    ast_value_vector.push_back(std::move(node_val));
+  }
+
+  for (size_t i = 7; i < vals.size(); ++i) {
+    auto node_val = tdb_unique_ptr<ASTNode>(tdb_new(
+        ASTNodeVal, "x", &vals[i], sizeof(vals[i]), QueryConditionOp::NE));
+    CHECK(
+        ast_node_to_str(node_val) ==
+        "x NE 0" + std::to_string(vals[i]) + " 00 00 00");
+    ast_value_vector.push_back(std::move(node_val));
+  }
+
+  int x = 6;
+  auto x_neq_six = tdb_unique_ptr<ASTNode>(
+      tdb_new(ASTNodeVal, "x", &x, sizeof(x), QueryConditionOp::NE));
+  CHECK(ast_node_to_str(x_neq_six) == "x NE 06 00 00 00");
+
+  auto one_or_two = ast_value_vector[0]->combine(
+      ast_value_vector[1]->clone(), QueryConditionCombinationOp::OR);
+  CHECK(
+      ast_node_to_str(one_or_two) == "(x EQ 01 00 00 00 OR x EQ 02 00 00 00)");
+
+  auto three_or_four = ast_value_vector[2]->combine(
+      ast_value_vector[3]->clone(), QueryConditionCombinationOp::OR);
+  CHECK(
+      ast_node_to_str(three_or_four) ==
+      "(x EQ 03 00 00 00 OR x EQ 04 00 00 00)");
+
+  auto six_or_seven = ast_value_vector[5]->combine(
+      ast_value_vector[6]->clone(), QueryConditionCombinationOp::OR);
+
+  CHECK(
+      ast_node_to_str(six_or_seven) ==
+      "(x EQ 06 00 00 00 OR x EQ 07 00 00 00)");
+
+  auto eight_and_nine = ast_value_vector[7]->combine(
+      ast_value_vector[8]->clone(), QueryConditionCombinationOp::AND);
+  CHECK(
+      ast_node_to_str(eight_and_nine) ==
+      "(x NE 08 00 00 00 AND x NE 09 00 00 00)");
+
+  auto subtree_a =
+      one_or_two->combine(three_or_four, QueryConditionCombinationOp::AND);
+  CHECK(
+      ast_node_to_str(subtree_a) ==
+      "((x EQ 01 00 00 00 OR x EQ 02 00 00 00) AND (x EQ 03 00 00 00 OR x EQ "
+      "04 00 00 00))");
+
+  auto subtree_d =
+      eight_and_nine->combine(six_or_seven, QueryConditionCombinationOp::AND);
+  CHECK(
+      ast_node_to_str(subtree_d) ==
+      "(x NE 08 00 00 00 AND x NE 09 00 00 00 AND (x EQ 06 00 00 00 OR x EQ 07 "
+      "00 00 00))");
+
+  auto subtree_c = subtree_d->combine(
+      ast_value_vector[4]->clone(), QueryConditionCombinationOp::OR);
+
+  CHECK(
+      ast_node_to_str(subtree_c) ==
+      "((x NE 08 00 00 00 AND x NE 09 00 00 00 AND (x EQ 06 00 00 00 OR x EQ "
+      "07 00 00 00)) OR x EQ 05 00 00 00)");
```

Comment:
Could we also test the case of an OR with both left-hand and right-hand AND clauses?  (Similarly in unit-QueryCondition).

Reply:
yes, this is resolved in the current pr